### PR TITLE
Improve code coverage

### PR DESCRIPTION
**Pull request by Mutombe**
## Summary

This pull request addresses issue #886 by introducing comprehensive tests for five views in the Projects app. These tests ensure the reliability and correctness of the views, enhancing the overall quality and stability of the application.

## Changes

The following changes have been made in this pull request:

1. Implemented tests for the `ProjectCreateAPIView`: These tests cover various scenarios related to project creation, validating inputs, handling permissions, and verifying the appropriate response codes and data.

2. Implemented tests for the `ProjectUpdateAPIView`: These tests thoroughly examine the functionality of updating project details, including edge cases, error handling, and ensuring the proper behavior of the view under different circumstances.

3. Implemented tests for the `ProjectDeleteAPIView`: These tests assess the behavior of the project deletion process, verifying that the appropriate permissions are enforced, confirming the correct response codes, and ensuring the database integrity is maintained.

4. Implemented tests for the `ProjectDetailsAPIView`: These tests comprehensively evaluate the functionality of retrieving project details, covering different scenarios, handling various input cases, and validating the correctness of the returned data.

Each of these test implementations  incorporates thorough test coverage to safeguard against regressions and ensure the robustness of the Projects app.

By introducing these tests, we can confidently assert the reliability and stability of these critical views, mitigating potential issues and providing a more professional and polished experience for our users.

Please review the changes and provide your feedback.

### CODE REVIEW

File: zubhub_backend/zubhub/creators/views.py [hunk 1]
```diff
@@ -134,7 +134,7 @@ def get_serializer_class(self):
             return CreatorMinimalSerializer
         else:
             return CreatorSerializer
-
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/creators/views.py:137]: Hello @Mutombe , thank you for working on this. You added extra space to this line. Can you remove it since it is not part of this PR?

File: zubhub_backend/zubhub/projects/models.py [hunk 2]
```diff
@@ -162,7 +162,7 @@ class Image(models.Model):
                                 null=True,
                                 related_name="images",
                                 blank=True)
-    image_url = models.URLField(max_length=1000)
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/models.py]: Here too, remove the extra space

File: zubhub_backend/zubhub/projects/serializers.py [hunk 3]
```diff
@@ -89,7 +89,7 @@ class CategorySerializer(serializers.ModelSerializer):
     class Meta:
         model = Category
         fields = [
-            "id",
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/serializers.py]: remove extra space

File: zubhub_backend/zubhub/projects/utils.py [hunk 4]
```diff
@@ -31,7 +31,7 @@ def task_lock(key):
 
 
 def update_images(project, images_data):
-    Image = apps.get_model('projects.Image')
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/utils.py]: remove extra space

File: zubhub_backend/zubhub/projects/views.py [hunk 5]
```diff
@@ -37,7 +37,7 @@ class ProjectCreateAPIView(CreateAPIView):
     Request body format:\n
         {\n
             "title": "string",\n
-            "description": "string",\n
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/views.py]: remove extra space

File: zubhub_frontend/zubhub/src/views/team/Team.jsx [hunk 6]
```diff
@@ -92,11 +92,11 @@ function Team(props) {
     drafts: [],
     badge_tags: [],
   });
-  const [followers,setFollowers]=React.useState([]);
+  const [followers, setFollowers]=React.useState([]);
   
   React.useEffect(() => {
   try{
-    let activitylogObj= new API()
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/views/team/Team.jsx]: remove extra space

File: zubhub_frontend/zubhub/src/views/team/Team.jsx [hunk 7]
```diff
@@ -92,11 +92,11 @@ function Team(props) {
     drafts: [],
     badge_tags: [],
   });
-  const [followers,setFollowers]=React.useState([]);
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/views/team/Team.jsx]: remove extra space

File: zubhub_frontend/zubhub/src/views/edit_team/EditTeam.jsx [hunk 8]
```diff
@@ -53,7 +53,7 @@ import {
   handleClickShowPassword,
   handleMouseDownPassword,
   getTeamProfile,
-} from './editTeamScripts';
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/views/edit_team/EditTeam.jsx]: remove extra space

File: zubhub_frontend/zubhub/src/store/actions/userActions.js [hunk 9]
```diff
@@ -155,7 +155,7 @@ export const getAllTeams = args => {
  */
 export const editTeam = args => {
   return dispatch => {
-    let profile;
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/store/actions/userActions.js]: remove extra space

File: zubhub_frontend/zubhub/src/components/form/videoInput/VideoInput.jsx [hunk 10]
```diff
@@ -12,7 +12,7 @@ import { videoInputStyles } from './videoInput.styles';
 function VideoInput({ name, label, required, value = [], handleChange, linkValue = '' }) {
   const commomClasses = makeStyles(styles)();
   const classes = makeStyles(videoInputStyles)();
-  const input = useRef(null);
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/components/form/videoInput/VideoInput.jsx]: remove extra space

File: zubhub_frontend/zubhub/src/App.js [hunk 11]
```diff
@@ -15,10 +15,24 @@ const SearchResults = React.lazy(() => import('./views/search_results/SearchResu
...
+  import('./views/email_confirm/EmailConfirm'),
+); 
+const PhoneConfirm = React.lazy(() =>
+  import('./views/phone_confirm/PhoneConfirm'),
+);
 const PasswordReset = React.lazy(() => import('./views/password_reset/PasswordReset'));
 const PasswordResetConfirm = React.lazy(() => import('./views/password_reset_confirm/PasswordResetConfirm'));
 const EmailConfirm = React.lazy(() => import('./views/email_confirm/EmailConfirm'));
 const PhoneConfirm = React.lazy(() => import('./views/phone_confirm/PhoneConfirm'));
+
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/App.js]: remove extra space

File: zubhub_frontend/zubhub/src/App.js [hunk 12]
```diff
@@ -15,10 +15,24 @@ const SearchResults = React.lazy(() => import('./views/search_results/SearchResu
...
+);
+const PasswordResetConfirm = React.lazy(() =>
+  import('./views/password_reset_confirm/PasswordResetConfirm'),
+);
+const EmailConfirm = React.lazy(() =>
+  import('./views/email_confirm/EmailConfirm'),
+); 
+const PhoneConfirm = React.lazy(() =>
+  import('./views/phone_confirm/PhoneConfirm'),
+);
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/App.js]: This change doesn't fit the description of the PR. If I am not mistaken this is likely from some kind of rebase/merge gone wrong. Do you mind removing this change if this is so? if this is indeed a part of this PR, do you mind changing the PR description to mention what these changes do?

File: zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx [hunk 13]
```diff
@@ -1,4 +1,5 @@
 import { Box, FormControl, IconButton, Typography, makeStyles } from '@material-ui/core';
+import { resizeImageFile } from 'react-image-file-resizer';
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx]: same merge/rebase issue

File: zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx [hunk 14]
```diff
@@ -63,14 +64,31 @@ export default function ImageInput({
     }
   };
 
-  const handleFileChange = files => {
+  const handleFileChange = async (files) => {
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx]: possibly same merge/rebase issue

File: zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx [hunk 15]
```diff
@@ -63,14 +64,31 @@ export default function ImageInput({
...
+      const file = files[i];
+      const compressedImage = await new Promise((resolve) => {
+        resizeImageFile(file, compressOptions.maxWidth, compressOptions.maxHeight, 'JPEG', compressOptions.quality, 0, (compressedFile) => {
+          resolve(compressedFile);
+        });
+      });
+      compressedImages.push(compressedImage);
+    }
+  
+    const images = [...compressedImages, ...value];
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx]: same merge/rebase issue

File: zubhub_frontend/zubhub/src/views/team/Team.jsx [hunk 16]
```diff
@@ -177,7 +177,10 @@ function Team(props) {
                     primaryButtonStyle
                     onClick={() => props.history.push('/'+groupname+'/edit-team')}
                   >
+                    {t('profile.editTeam')} 
+
                     {t('editTeam.info')}
+
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/views/team/Team.jsx]: same merge/rebase issue

File: zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx [hunk 17]
```diff
@@ -13,7 +14,7 @@ const defaultMessage = 'JPG and PNG Images can be added (Maximum of 5 photos 2MB
 export default function ImageInput({
   name,
   label,
-  required,
+  required,
```
Comment (NdibeRaymond) [inline zubhub_frontend/zubhub/src/components/form/imageInput/ImageInput.jsx]: remove extra space

**Review by NdibeRaymond — CHANGES_REQUESTED** [review]
Hello @Mutombe, thanks for much for working on this! testing is something we are really lacking and I am glad someone is taking on that. I requested a couple of changes. Do you mind taking a look?

File: zubhub_backend/zubhub/projects/tests.py [hunk 18]
```diff
@@ -1,3 +1,139 @@
...
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+        # Add more assertions for the response data
+
+    # Add more test cases for other scenarios
+
+
+class ProjectUpdateAPITest(TestCase):
+    def setUp(self):
+        self.factory = RequestFactory()
+        self.project = Project.objects.create(title='Test Project')
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/tests.py]: I think we should make the project belong to a particular user. That way we can test if a user can edit another user's project.

File: zubhub_backend/zubhub/projects/tests.py [hunk 19]
```diff
@@ -1,3 +1,139 @@
...
+        request = self.factory.patch(f'/projects/{self.project.pk}/', data)
+        request.user = AnonymousUser()
+
+        response = ProjectUpdateAPIView.as_view()(request, pk=self.project.pk)
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+class ProjectDeleteAPITest(TestCase):
+    def setUp(self):
+        self.factory = RequestFactory()
+        self.project = Project.objects.create(title='Test Project')
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/tests.py]: I think we should make the project belong to a particular user. That way we can test if a user can delete another user's project.

File: zubhub_backend/zubhub/projects/tests.py [hunk 20]
```diff
@@ -1,3 +1,139 @@
...
+        request.user = AnonymousUser()
+
+        response = ProjectDeleteAPIView.as_view()(request, pk=self.project.pk)
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
+
+
+class ProjectDetailsAPITest(TestCase):
+    def setUp(self):
+        self.factory = RequestFactory()
+        self.project = Project.objects.create(title='Test Project')
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/tests.py]: I think we should make the project belong to a particular user. That way we can test if a user can  see details about another user's project they are not allowed to see.

File: zubhub_backend/zubhub/projects/tests.py [hunk 21]
```diff
@@ -1,3 +1,139 @@
...
+
+    def test_get_project_details_authenticated_user(self):
```
Comment (NdibeRaymond) [inline zubhub_backend/zubhub/projects/tests.py:129]: we can try inspecting the response data to ensure the user can only see the details they are allowed to see (both for authenticated and unauthenticated)

File: zubhub_backend/zubhub/projects/models.py (hunk 2 above)
Comment (Mutombe) [reply zubhub_backend/zubhub/projects/models.py]: Done

File: zubhub_backend/zubhub/projects/serializers.py (hunk 3 above)
Comment (Mutombe) [reply zubhub_backend/zubhub/projects/serializers.py]: Done

File: zubhub_backend/zubhub/projects/utils.py (hunk 4 above)
Comment (Mutombe) [reply zubhub_backend/zubhub/projects/utils.py]: Done

File: zubhub_backend/zubhub/projects/views.py (hunk 5 above)
Comment (Mutombe) [reply zubhub_backend/zubhub/projects/views.py]: Done

File: zubhub_frontend/zubhub/src/views/team/Team.jsx (hunk 6 above)
Comment (Mutombe) [reply zubhub_frontend/zubhub/src/views/team/Team.jsx]: Done

File: zubhub_frontend/zubhub/src/views/team/Team.jsx (hunk 7 above)
Comment (Mutombe) [reply zubhub_frontend/zubhub/src/views/team/Team.jsx]: Done

**Comment by Mutombe** [comment]
> Hello @Mutombe, thanks for much for working on this! testing is something we are really lacking and I am glad someone is taking on that. I requested a couple of changes. Do you mind taking a look?

Roger that.

File: zubhub_backend/zubhub/projects/tests.py (hunk 18 above)
Comment (Mutombe) [reply zubhub_backend/zubhub/projects/tests.py]: You're right! By associating the project with a specific user, you can test the behavior of users trying to edit projects that don't belong to them.

File: zubhub_backend/zubhub/projects/tests.py (hunk 20 above)
Comment (Mutombe) [reply zubhub_backend/zubhub/projects/tests.py]: I have made some changes

**Comment by NdibeRaymond** [comment]
hello @Mutombe, I see that you still have have of the things in the previous review in this patch. I'd advise you remove them so we can merge this. I can work you  through the process of doing this if you'd like:
* first update your master branch. this command `git checkout master && git pull origin master` will do that for you.
* next create a new branch with `git checkout -b name_of_the_branch_here`
* next copy over your main changes from `2nd-branch` to the new branch. you might have to do this manually
* commit the changes and push to github.
* then you have to change the base branch of this PR to the new branch. If that is difficult you might need to create a new PR and close this one.
* also when next you want to merge changes from the master branch into your working branch, use `rebase` instead of `pull` as this helps to keep the history clean and makes review easy. to do this you start with `git rebase -i origin/master` and follow the provided instructions. you can also read this online and it probably won't take more than 20mins to get the hang of it

**Comment by NdibeRaymond** [comment]
also ensure that your own copy of the repository is up to date

**Comment by Mutombe** [comment]
Done

https://github.com/unstructuredstudio/zubhub/pull/987
